lib/model/model.go

```diff
@@ -775,25 +781,49 @@ func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset
 			}
 			return protocol.ErrGeneric
 		}
-		reader = strings.NewReader(target)
-	} else {
-		// Cannot easily cache fd's because we might need to delete the file
-		// at any moment.
-		reader, err = os.Open(fn)
-		if err != nil {
-			l.Debugln("os.Open:", err)
-			if os.IsNotExist(err) {
-				return protocol.ErrNoSuchFile
-			}
+		if _, err := strings.NewReader(target).ReadAt(buf, offset); err != nil {
+			l.Debugln("symlink.Reader.ReadAt", err)
 			return protocol.ErrGeneric
 		}
+		return nil
+	}
+
+	// Cannot easily cache fd's because we might need to delete the file
+	// at any moment.
+
+	// Only check temp files if we the flag is set, and if we are set to
+	// advertise the temp indexes.
+	if flags&protocol.FlagFromTemporary == protocol.FlagFromTemporary && m.cfg.Options().SendTempIndexes {
+		tempFn := filepath.Join(folderPath, defTempNamer.TempName(name))
+		tempFile, err := os.Open(tempFn)
+		if err == nil {
+			defer tempFile.Close()
+			_, err = tempFile.ReadAt(buf, offset)
+			if err != nil {
+				// The file might have finishe downloading, try the real file.
```

Comment:
I think this part is a bit muddled, I was staring for too long at the err==nil/err!=nil logic, as it differs from the usual and couldn't just be processed automatically. :) How about breaking out functions `readFromTemporary`, `readFromCurrent` to read into buf or return an error or something to simplify this?